Weekly_Work/Week1/Code_changes-git/Metrics_Root_class.py

```diff
@@ -0,0 +1,202 @@
+                    clean_pr_list.append(clean_line)
+                        
+                        
+                self.clean_commit_df = pd.DataFrame(clean_commit_list)
+                self.clean_issue_df = pd.DataFrame(clean_issue_list)
+                self.clean_pr_df = pd.DataFrame(clean_pr_list)
+                
+            self.clean_dict = {
+                'commit': self.clean_commit_df,
+                'issue': self.clean_issue_df,
+                'pull_request': self.clean_pr_df
+            }
+       
+        if self.since:
+            for df in self.clean_dict.values():
+                df = df[df['created_date'] >= self._str_to_dt_other(self.since)]
+        else: 
+            self.since = self._get_date("since")
+            
+        if self.until:
+            for df in self.clean_dict.values():
+                df = df[df['created_date'] < self._str_to_dt_other(self.until)]
+        else: 
+            self.until = self._get_date("until")
+        
+    def _clean_commit(self, line):
+        cdata = line['data']
+        cleaned_line =  \
+        {
+            'repo': line['origin'],
+            'hash': cdata['commit'],
+            'author': cdata['Author'],
+            'category': "commit",
+            'created_date': self._str_to_dt_data(cdata['AuthorDate']),
+            'commit': cdata['Commit'],
+            'commit_date': self._str_to_dt_data(cdata['CommitDate']),
+            'files_no': len(cdata['files']),
+            'refs': cdata['refs'],
+            'parents': cdata['parents'],
+            'files': cdata['files']
+        }
+
+        actions = 0
+        for file in cdata['files']:
+            if 'action' in file:
+                actions += 1
+        cleaned_line['files_action'] = actions
+        if 'Merge' in cdata:
+            cleaned_line['merge'] = True
+        else:
+            cleaned_line['merge'] = False
+
+        return cleaned_line
+            
+    def _clean_issue(self, line):
+        repo_name = line['origin']
+        line_data = line['data']
+        cleaned_line =  \
+        {
+            'repo': repo_name,
+            'hash': line_data['id'],
+            'category': "issue",
+            'author': line_data['user']['login'],
+            'created_date': self._str_to_dt_data(line_data['created_at']),
+            'current_status': line_data['state']   
+        }
+        return cleaned_line
+
+            
+    def _clean_pr(self, line):
+        repo_name = line['origin']
+        line_data = line['data']
+        cleaned_line =  \
+        {
+            'repo': repo_name,
+            'hash': line_data['id'],
+            'category': "pull_request",
+            'author': line_data['user']['login'],
+            'created_date': self._str_to_dt_data(line_data['created_at']),
+            'current_status': line_data['state']   
+        }
+        return cleaned_line
+
+    def _is_source_code(self, commit):
```

Comment:
This should be a parameter of the instantiation: a function that, given a commit, would decide if it is related to source code or not. With the implementation of the metrics, we can provide some. The easiest one would just return "True" for all commits (naive assumption that everything is code).

Reply:
If I understood correctly, this would require several functions, since we are passing the functions directly as arguments for instantiation. What I have done is passed a list of file extensions to exclude through the variable `source_code_exclude_list`. The `_get_extension` function returns the extension of a file. The `is_source_code` method compares the file extensions of all files in a particular commit to the `source_code_exclude_list` and if all of the files in a commit are to be excluded, that commit is not considered for analysis.
The naive case, as you mentioned, happens when `source_code_exclude_list` is `None` and _is_source_code` returns True.
```python
if self.source_code_exclude_list is None or len(extension_set.difference(self.source_code_exclude_list)) > 0:
        return True
    return False
```

Could you please explain what you suggest? How will you include the file extensions to be excluded?

Reply:
What I mean is that the criteria for deciding what is code could be based in something else different from the file extension. For example, we could consider that tests are not "code", and that cannot be checked by looking at the extension of the files. I find the approach of using a function to filter the commits to consider as a generalization of your idea, and has a direct mapping with the idea in the metrics definition that "we should define what is source code": the function, whichever it is, would define what is source code.

So, for cases where you can detect source code by extension, there is really no difference (maybe I would still like the functional solution, but that would be maybe a personal preference). But if you think of more elaborate criteria for deciding what is source code, the functional approach can be extended as much as needed...

(Similar results could be obtained with a class instead of a function, but I think in this cases that would overcomplicate the case).